frame/support/src/traits/hooks.rs

```diff
@@ -239,64 +209,112 @@ impl OnRuntimeUpgrade for Tuple {
 	}
 }
 
-/// Type that provide some integrity tests.
-///
-/// This implemented for modules by `decl_module`.
+/// See [`Hooks::integrity_test`].
 #[cfg_attr(all(not(feature = "tuples-96"), not(feature = "tuples-128")), impl_for_tuples(64))]
 #[cfg_attr(all(feature = "tuples-96", not(feature = "tuples-128")), impl_for_tuples(96))]
 #[cfg_attr(feature = "tuples-128", impl_for_tuples(128))]
 pub trait IntegrityTest {
-	/// Run integrity test.
-	///
-	/// The test is not executed in a externalities provided environment.
+	/// See [`Hooks::integrity_test`].
 	fn integrity_test() {}
 }
 
-/// The pallet hooks trait. Implementing this lets you express some logic to execute.
+/// The pallet hooks trait. This is merely an umbrella trait for:
+///
+/// - [`OnInitialize`]
+/// - [`OnFinalize`]
+/// - [`OnRuntimeUpgrade`]
+/// - [`OffchainWorker`]
+/// - [`OnIdle`]
+/// - [`IntegrityTest`]
+///
+/// ## Ordering
+///
+/// For all hooks, except [`OnIdle`] the order of execution is derived from how the pallets are
+/// ordered in [`crate::construct_runtime`].
+///
+/// ## Summary
+///
+/// In short, the following diagram shows the
 pub trait Hooks<BlockNumber> {
-	/// The block is being finalized. Implement to have something happen.
+	/// Block initialization hook. This is called at the very beginning of the block execution
+	///
+	/// Must return the non-negotiable weight of both itself, and whatever [`on_finalize`] wishes to
+	/// consumed.
+	///
+	/// The weight returned by this is treated as [`DispatchClass::Mandatory`], meaning that it MUST
+	/// BE EXECUTED. This implies that this hook should only ever be used if the logic in it must
+	/// happen under any circumstance. For all similar use cases where a delay is acceptable, use
+	/// [`poll`].
+	///
+	/// NOTE: This function is called BEFORE ANY extrinsic in a block is applied, including inherent
+	/// extrinsics. Hence for instance, if you runtime includes `pallet-timestamp`, the `timestamp`
+	/// is not yet up to date at this point.
+	fn on_initialize(_n: BlockNumber) -> Weight {
+		Weight::zero()
+	}
+
+	/// Similar to [`on_initialize`], but:
+	///
+	/// 1. is called after [`on_initialize`] of all pallets is done executing,
+	/// 2. execution on *each block* is not guaranteed
+	///
+	/// The latter point is usually correlated with the block resources being available, for example
+	/// `Weight` leftover.
+	fn poll(_n: BlockNumber) -> Weight {
```

Comment:
should it take the remaining weight into argument, similarly to `on_idle` ?

Reply:
How exactly does this work? This is skipped when on_initialize consumes all the weight? This sounds like on_idle but before extrinsics.

Also the name is not consistent with all other hooks which are `on_xxx`

Reply:
`on_initialize` and `on_finalize` will be [deprecated](https://github.com/paritytech/substrate/issues/13690). `on_idle()` happens at the end of a block and only if it has weight left. `poll()` happens at the beginning, prior to non-mandatory transaction inclusion.

So:

- `initialize_block`
  - Pallets `on_initialize` code, though this either be migrated to `poll` or moved into a hard-deadline hook, called by a explicit configuration of `System` pallet.
- If no migrations scheduled or on-going: `poll`
- Mandatory extrinsics (inherents)
- Multi-block migrations
- If no migrations on-going: other extrinsics/transactions
- If weight left `on_idle`
- `finalize_block`
  - Pallets `on_finalize` code, though this either be migrated to `poll` or moved into a hard-deadline hook, called by a explicit configuration of `System` pallet.

Reply:
while we are here, what’s the plan for post inherent hook?

Reply:
Not sure about this - got a link?

Reply:
https://github.com/paritytech/substrate/issues/9210 https://github.com/paritytech/substrate/pull/10128

Reply:
AFAICT if we implement multi-block-migration after inherents then implementing a hook after inherents should be straightforward.

Reply:
Fair enough;

- `initialize_block`
  - Pallets `on_initialize` code, to be migrated to `poll` as much as possible.
- Inherents
- Multi-block migrations
- If no migrations on-going:
  - `poll`
  - other extrinsics/transactions
- If no migrations on-going and weight left `on_idle`
- `finalize_block`
  - Pallets `on_finalize` code, to be migrated to `poll` as much as possible.

However I'd feel better is all `on_initialize` and `on_finalize` hooks were moved over to a much more explicit System pallet configuration, using new items `frame_system::Config::PreInherents`, `::PostInherents` and `::PostTransactions` because Pallets should generally not be using these hooks since they not only make MBMs unsafe but they also make MBMs of any pallets which they call into unsafe.

The block-execution process would then become:

- `frame_system::Config::PreInherents::execute()`
- Inherents
- `frame_system::Config::PostInherents::execute()`
- Multi-block migrations
- If no migrations on-going:
  - `poll`
  - other extrinsics/transactions
- If no migrations on-going and weight left `on_idle`
- `frame_system::Config::PostTransactions::execute()`

Now this overall design makes me fell a little uncomfortable since it is unclear exactly at which point in the block initialisation process an extrinsic may be executed which seems like a recipe for trouble. We also have the open question about the `parachains_inherent` and whether it's really mandatory; if it is then it would imply that `pallet_message_queue` could never have an MBM in the model since it is called into be a ME. We might need to split the inherent into two parts in that case; one which is mandatory but which doesn't contain any messages and a second which is optional (and would not be included in the case of an MBM) and only does the message transport.

Reply:
> should it take the remaining weight into argument, similarly to on_idle ?

I considered this and am still open to it. End of the day, the developer can always query remaining weight at the beginning of the code and make an informed choice not to overfill the block. 

I have less of a strong opinion about which to chose, but more so to unify them. In all `Hooks` functions `block_number` and `remaining_weight` should be either passed in for convenience, or not passed in based on the assumption that you can always read them from `systme`.

Reply:
What would happen if there is a faulty MBM at hand, and it never stops? should MBMs have a deadline, after which we (partially) resume the chain? Ideally we want few governance calls to be resumed, but not everything else.

Reply:
Side note, I wonder what poeple think about my comment [here](https://github.com/paritytech/substrate/issues/13731#issuecomment-1582123883). I sometimes find such discussions to be confusing because we use the term `DispatchClass` even for things that are not _dispatchable_ (like `Hooks`), to my best knowledge. Moreover, something like `OnIdle` is declared `Mandatory`, even thought it is clearly not.  

https://github.com/paritytech/substrate/blob/b72c475803947121b3b475eb33ba5fc48f1fe5b6/frame/executive/src/lib.rs#L536

We can possibly formulate a more comprehensive (and ideally extensible) `enum WeightClass`, the variants of which encapsulate different weights that may or may not be permitted to be consumed. For example, each variant of this enum is encoded at a custom `codec_index(x)`, where `x` demonstrates its "priority". 

- MBMs are then configured to block all weights classes with "priority" less then some `y`. 
- Executive/System is configured to never permit over-consumption of weights of certain classes (extrinsics), but permit others to exceed their limits (hooks, perhaps except `on_idle`). 

Things like `frame_system::Config::PreInherents` or `on_initialize` are set to `codec_index(255)` which by definition means they cannot be blocked. 

Possibly, one could argue any code that has priority other than 255 needs to be benchmarked and be able to predict its weight consumption prior to execution. This is something that we have a soft-expectation about in `on_idle` now, but don't and cannot enforce. 

This also allows things like "permit sudo or governance calls during MBMs" to be expressed. 

Possibly an over-engineering, but I feel like this can:

1. Improve `pol`+`MBM`
2. close https://github.com/paritytech/substrate/issues/13731
3. close https://github.com/paritytech/substrate/issues/11309 

all at once.

Reply:
> We also have the open question about the `parachains_inherent` and whether it's really mandatory; if it is then it would imply that `pallet_message_queue` could never have an MBM in the model since it is called into be a ME. We might need to split the inherent into two parts in that case; one which is mandatory but which doesn't contain any messages and a second which is optional (and would not be included in the case of an MBM) and only does the message transport.

The messages could stay part of the inherent? However, when there is a MBM ongoing, we would need to buffer the inbound messages by default. Writing to the state can be done in the new format (assuming there would be some MBM running for `pallet-message-queue`). It would be good if we could announce to other chains that the message queue is full and we can not process more messages right now.

In general, we will always require that parts of the state are always migrated in one go. If for example `pallet-message-queue` has some counter for stored messages, we need to migrate this counter directly when starting the MBM. This can not be done later. I think this pattern of having certain values requiring an instant migration will be found for other pallets as well. Another example would be the list of authorities that should be migrated, which would probably also be required to be done at the start of a potential MBM.